### PR TITLE
Implement replacement for template path constants using the functions within in-memory cache in globals

### PR DESCRIPTION
**Pull request by felixarntz**
This PR is another attempt at replacing use of the `STYLESHEETPATH` and `TEMPLATEPATH` constants to aid testability with a hopefully minimal (as in negligible) performance impact.

It is a combination of the ideas explored in #4880 and #5156.

Trac ticket: https://core.trac.wordpress.org/ticket/18298

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by felixarntz** [comment]
Running benchmarks of this doesn't show a notable difference either way:
* I used `npm run research -- benchmark-server-timing -u http://localhost:8889 -n 100`, and ran it twice for every scenario.

### Twenty Twenty-Three results
1. 93.31ms with PR vs 96.21ms without PR
2. 98.97ms with PR vs 97.16ms without PR

### Twenty Twenty-One results
1. 57.15ms with PR vs 58.25ms without PR
2. 57.21ms with PR vs 56.81ms without PR

Based on this data, it seems that the difference in performance is tiny enough to not care.

**Comment by felixarntz** [comment]
@kt-12 Would be great to get your thoughts on this as well.

File: src/wp-includes/theme.php [hunk 1]
```diff
@@ -832,6 +853,13 @@ function switch_theme( $stylesheet ) {
 
 	update_option( 'theme_switched', $old_theme->get_stylesheet() );
 
+	/*
+	 * Reset globals to force refresh the next time these directories are
+	 * accessed via `get_stylesheet_directory()` / `get_template_directory()`.
+	 */
+	$wp_stylesheet_path = null;
+	$wp_template_path   = null;
+
```
Comment (kt-12) [inline src/wp-includes/theme.php:875]: @felixarntz The creation of `wp_stylesheet_path` and `wp_template_path` global will lead to a slim chance of people using the global directly rather that via the function (as we can't force them to stick to `functions only`, using some notice or error message)

A sites depending extensively on global  will get error after they `switch_theme`, as both the path globals are going to be null. I would prefer us updating the global immediately after `switch_theme`

**Review by kt-12 — CHANGES_REQUESTED** [review]
Thank you @felixarntz. Your version is much neater than other two. I have left one comment ([here](https://github.com/WordPress/wordpress-develop/pull/5160#discussion_r1318459853)).

I'll focus on moving test cases from #4880 here. However during #4880 tests I did hit a roadblocker with the `theme.php` test file due to the usage of `wp_clean_themes_cache()` in setup. I need to check if that is still an issue once I move the code here.

File: src/wp-includes/theme.php (hunk 1 above)
Comment (felixarntz) [reply src/wp-includes/theme.php:875]: @kt-12 IMO that's not something to be concerned of. WordPress uses lots of globals, and generally those globals should never be used directly, except for a few exceptions such as `$wpdb`.

We could maybe rename them to use an underscore prefix to clarify via code, i.e. `$_wp_*`?

**Comment by felixarntz** [comment]
@kt-12 I have migrated some of your tests from #4880 over to this PR and added a few more tests. I think the issues in your tests with the theme root were a specific issue with the way the test setup was handled; most importantly, using filters to override the theme root should not be necessary here, or rather would not be related to the change that is being tested here.

I've also updated the existing (very old) `test_switch_theme()` test specifically to "un-comment" the commented out code for the `get_query_template()` logic which used to be not testable but now is - so that's a neat way to ensure the change here works. Obviously that test is very messy and huge, but rather than bothering to clean it up, I think it can serve as a good demonstration that what was intended years ago now works that way finally.

Of course I've also added a few more targeted tests, to make sure we also have coverage that reflects our current testing standards :)

File: src/wp-includes/default-constants.php [hunk 2]
```diff
@@ -407,13 +407,15 @@ function wp_templating_constants() {
 	 * Filesystem path to the current active template directory.
 	 *
 	 * @since 1.5.0
+	 * @deprecated 6.4.0 Use `get_template_directory()` instead.
```
Comment (mukeshpanchal27) [inline src/wp-includes/default-constants.php]: It's important to improve the documentation for this function by adding a `@see` annotation, similar to how it's done for other `deprecated` functions.

There are two options available, but the majority of other `deprecated` functions use the first option:

1.
```suggestion
	 * @deprecated 6.4.0 Use {@see 'get_template_directory()'} instead.
```

2.
```suggestion
	 * @deprecated 6.4.0 Use get_template_directory() instead.
	 * @see get_template_directory()
```


🔢 If changed, applies elsewhere in the PR.

File: src/wp-includes/theme.php [hunk 3]
```diff
@@ -157,7 +157,10 @@ function wp_clean_themes_cache( $clear_update_cache = true ) {
  * @return bool True if a child theme is in use, false otherwise.
  */
 function is_child_theme() {
-	return ( TEMPLATEPATH !== STYLESHEETPATH );
+	$stylesheet_path = get_stylesheet_directory();
+	$template_path   = get_template_directory();
+
+	return ( $template_path !== $stylesheet_path );
```
Comment (mukeshpanchal27) [inline src/wp-includes/theme.php]:
```suggestion
	return ( get_template_directory() !== get_stylesheet_directory() );
```
I noticed that you have two variables, `$stylesheet_path` and `$template_path`, which are only used once in the code. Given that they are used only once, we can simplify the code by directly comparing the functions `get_template_directory()` and `get_stylesheet_directory()`. This will make the code more concise and easier to read.

However, if there are other purposes for these variables that I'm not aware of, please let me know.

File: src/wp-includes/theme.php [hunk 4]
```diff
@@ -188,23 +191,31 @@ function get_stylesheet() {
  *
  * @since 1.5.0
  *
+ * @global string $wp_stylesheet_path Current theme stylesheet directory path.
```
Comment (mukeshpanchal27) [inline src/wp-includes/theme.php:192]: Cleverly manage global variables in the function. Awesome work, Felix.

Do we need to add/update documentation for the function so others know that starting from version 6.4, we implemented these changes? If not, then mark it as resolved.

🔢 If changed, applies elsewhere in the PR.

File: src/wp-includes/theme.php (hunk 1 above)
Comment (mukeshpanchal27) [reply src/wp-includes/theme.php:875]: I agree with @felixarntz's thought here. If people are getting results from the function, they might not use global variables, but we can't restrict them from using globals.

File: tests/phpunit/tests/template.php [hunk 5]
```diff
@@ -629,6 +629,42 @@ public function data_extra_args_for_add_settings_section() {
 		);
 	}
 
+	/**
+	 * Tests that `locate_template()` uses the current theme even after switching the theme.
+	 *
+	 * @ticket 18298
+	 * @covers ::locate_template
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/template.php]:
```suggestion
	 * @ticket 18298
	 *
	 * @covers ::locate_template
```
🔢 If changed, applies elsewhere in the PR.

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks, @felixarntz, for the PR. I've left some initial feedback for your consideration. Overall, the idea looks good to me.

**Comment by mukeshpanchal27** [comment]
The https://3v4l.org/D2WrI, demonstrates that the performance impact is negligible and not a significant concern.

File: src/wp-includes/theme.php (hunk 1 above)
Comment (kt-12) [reply src/wp-includes/theme.php:875]: @felixarntz I don't think we should rename it to `$_wp_*`. I was thinking if we could reinitialise global immediately after we `switch_theme` the value obtained from usage of global will also be true at all given points. 
Anyway, it's a very rare scenario(unless proven otherwise in future) that someone would opt for global instead of the function so we can leave it the way it is.

**Review by kt-12 — APPROVED** [review]
Except for changes request by @mukeshpanchal27, everything else LGTM!. Thanks @felixarntz.

File: src/wp-includes/template.php [hunk 6]
```diff
@@ -699,16 +700,19 @@ function get_attachment_template() {
...
 			continue;
 		}
-		if ( file_exists( STYLESHEETPATH . '/' . $template_name ) ) {
-			$located = STYLESHEETPATH . '/' . $template_name;
+		if ( file_exists( $stylesheet_path . '/' . $template_name ) ) {
+			$located = $stylesheet_path . '/' . $template_name;
 			break;
-		} elseif ( is_child_theme() && file_exists( TEMPLATEPATH . '/' . $template_name ) ) {
-			$located = TEMPLATEPATH . '/' . $template_name;
+		} elseif ( is_child_theme() && file_exists( $template_path . '/' . $template_name ) ) {
```
Comment (spacedmonkey) [inline src/wp-includes/template.php]:
```suggestion
		} elseif ( $is_child_theme && file_exists( $template_path . '/' . $template_name ) ) {
```

File: src/wp-includes/theme.php [hunk 7]
```diff
@@ -188,23 +191,31 @@ function get_stylesheet() {
...
+		/**
+		 * Filters the stylesheet directory path for the active theme.
+		 *
+		 * @since 1.5.0
+		 *
+		 * @param string $stylesheet_dir Absolute path to the active theme.
+		 * @param string $stylesheet     Directory name of the active theme.
+		 * @param string $theme_root     Absolute path to themes directory.
+		 */
+		$wp_stylesheet_path = apply_filters( 'stylesheet_directory', $stylesheet_dir, $stylesheet, $theme_root );
```
Comment (spacedmonkey) [inline src/wp-includes/theme.php]: Using a global like this mean that the filters will only fire once. This maybe a breaking change.

File: src/wp-includes/theme.php [hunk 8]
```diff
@@ -322,23 +333,31 @@ function get_template() {
...
+		/**
+		 * Filters the active theme directory path.
+		 *
+		 * @since 1.5.0
+		 *
+		 * @param string $template_dir The path of the active theme directory.
+		 * @param string $template     Directory name of the active theme.
+		 * @param string $theme_root   Absolute path to the themes directory.
+		 */
+		$wp_template_path = apply_filters( 'template_directory', $template_dir, $template, $theme_root );
```
Comment (spacedmonkey) [inline src/wp-includes/theme.php]: See https://github.com/WordPress/wordpress-develop/pull/5160/files#r1319822198

File: src/wp-includes/template.php [hunk 9]
```diff
@@ -699,16 +700,19 @@ function get_attachment_template() {
  * @return string The template filename if one is located.
  */
 function locate_template( $template_names, $load = false, $load_once = true, $args = array() ) {
+	$stylesheet_path = get_stylesheet_directory();
+	$template_path   = get_template_directory();
+
```
Comment (spacedmonkey) [inline src/wp-includes/template.php:706]:
```suggestion
     $is_child_theme = $template_path !== $stylesheet_path;
```

**Comment by spacedmonkey** [comment]
I am seeing an almost 3% improvement. https://blackfire.io/profiles/compare/94000d73-ac1d-4149-8db9-b034d9fb1682/graph

File: src/wp-includes/default-constants.php (hunk 2 above)
Comment (felixarntz) [reply src/wp-includes/default-constants.php]: @mukeshpanchal27 Will update to the 2nd version, that's the correct one looking at other code code. We shouldn't use `@see` within the `@deprecated` annotation.

File: src/wp-includes/theme.php (hunk 7 above)
Comment (felixarntz) [reply src/wp-includes/theme.php]: Correct, the filter will fire once per page load, except when the active theme has changed since the last time (because of the global being reset).

I agree with you that there is a small risk here, but using the filter to return a different theme directory for a single theme being active is very unlikely. I think we should post a draft note to explain this change (same for the `template_directory` filter) in case someone is using it in weird ways. It's still possible to force the filter to re-run by manually setting the global back to `null`. Obviously not recommended in general, but if there's a very specific use-case, it would be possible.

File: src/wp-includes/theme.php (hunk 8 above)
Comment (felixarntz) [reply src/wp-includes/theme.php]: See my reply in https://github.com/WordPress/wordpress-develop/pull/5160#discussion_r1320186178, I think we can keep this as is, and mark the ticket to need a dev note.

File: src/wp-includes/default-constants.php (hunk 2 above)
Comment (felixarntz) [reply src/wp-includes/default-constants.php]: Addressed in https://github.com/WordPress/wordpress-develop/pull/5160/commits/03f884431bdae8f49d4aadde6ac9587239a84434

File: src/wp-includes/theme.php (hunk 3 above)
Comment (felixarntz) [reply src/wp-includes/theme.php]: Addressed in https://github.com/WordPress/wordpress-develop/pull/5160/commits/a2b557418c4a47dc776dedc246eb8473bf23ea8f

File: src/wp-includes/theme.php (hunk 4 above)
Comment (felixarntz) [reply src/wp-includes/theme.php:192]: Addressed in https://github.com/WordPress/wordpress-develop/pull/5160/commits/96b5f650833c15f1a97931b5a7d5a52e14201c3a

File: tests/phpunit/tests/template.php (hunk 5 above)
Comment (felixarntz) [reply tests/phpunit/tests/template.php]: Addressed in https://github.com/WordPress/wordpress-develop/pull/5160/commits/7bc339ec9f14f53b1e9777bbc01a1200a760a959

File: src/wp-includes/template.php (hunk 6 above)
Comment (felixarntz) [reply src/wp-includes/template.php]: Addressed in https://github.com/WordPress/wordpress-develop/pull/5160/commits/928936f22dcf2bb0c16e98aaa910ce65dfba1780

File: src/wp-includes/template.php (hunk 9 above)
Comment (felixarntz) [reply src/wp-includes/template.php:706]: Addressed in https://github.com/WordPress/wordpress-develop/pull/5160/commits/928936f22dcf2bb0c16e98aaa910ce65dfba1780

**Comment by felixarntz** [comment]
@mukeshpanchal27 @spacedmonkey I have addressed all your feedback. Please take another look.

@spacedmonkey Thank you for the profiling! I think the 3% improvement confirms that there is no negligible performance impact. I don't think the 3% are an _actual_ performance improvement, I assume they just come from variance, similarly to how in my own benchmarks sometimes the PR was faster, sometimes slower, but always just a tiny bit.

Since we're replacing globals with function calls, I don't think in reality there is any chance that the code here is faster than the one before. The point is mostly to ensure that it doesn't become slower to a notable degree - which I think we have proven with the different benchmarks we've been running against this. 👍

File: src/wp-includes/template.php [hunk 10]
```diff
@@ -699,16 +700,20 @@ function get_attachment_template() {
  * @return string The template filename if one is located.
  */
 function locate_template( $template_names, $load = false, $load_once = true, $args = array() ) {
+	$stylesheet_path = get_stylesheet_directory();
+	$template_path   = get_template_directory();
+	$is_child_theme  = is_child_theme();
```
Comment (mukeshpanchal27) [inline src/wp-includes/template.php]:
```suggestion
	$is_child_theme  = $template_path !== $stylesheet_path;
```

The `is_child_theme()` function calls the `get_template_directory()` and `get_stylesheet_directory()` functions again, so let's utilize the already defined local variables, as Jonny also suggested in https://github.com/WordPress/wordpress-develop/pull/5160#discussion_r1319825399.

File: tests/phpunit/tests/template.php [hunk 11]
```diff
@@ -629,6 +629,43 @@ public function data_extra_args_for_add_settings_section() {
...
+		// Look for parent themes with an index.php template.
+		$themes = array_values(
+			array_filter(
+				wp_get_themes(),
+				static function( $theme ) {
+					if ( $theme->get_stylesheet() !== $theme->get_template() ) {
+						return false;
+					}
+					$php_templates = $theme['Template Files'];
+					return isset( $php_templates['index.php'] );
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/template.php]: Are these changes made for backward compatibility?

File: tests/phpunit/tests/general/template.php [hunk 12]
```diff
@@ -41,6 +42,7 @@ public static function wpTearDownAfterClass() {
 	public function set_up() {
 		parent::set_up();
 
+		switch_theme( 'default' );
```
Comment (mukeshpanchal27) [inline tests/phpunit/tests/general/template.php:45]:
```suggestion
		switch_theme( WP_DEFAULT_THEME );
```

Use constant here. 

🔢 Applies to a others too

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks to @felixarntz for the changes. The PR looks good to go, but I've left some feedback and questions for consideration.

File: tests/phpunit/tests/template.php (hunk 11 above)
Comment (felixarntz) [reply tests/phpunit/tests/template.php]: @mukeshpanchal27 See the comment above, this is to only consider themes with an `index.php` template (i.e. not a block theme).

For block themes without any PHP templates, the results may differ so that's not something this test could cover.

File: tests/phpunit/tests/general/template.php (hunk 12 above)
Comment (felixarntz) [reply tests/phpunit/tests/general/template.php:45]: @mukeshpanchal27 Not sure about this. The `default` is hard-coded in several unit tests, and the `WP_DEFAULT_THEME` constant may have a different value. So it's safer to guarantee that `default` is used.

File: src/wp-includes/template.php (hunk 10 above)
Comment (felixarntz) [reply src/wp-includes/template.php]: Updated in https://github.com/WordPress/wordpress-develop/pull/5160/commits/b9eb49db7a8af7bef0931d35100f88ef019ab9ef

File: src/wp-includes/theme.php (hunk 7 above)
Comment (spacedmonkey) [reply src/wp-includes/theme.php]: Might want to tests again some of these plugins 

https://www.wpdirectory.net/search/01HA2CHSDJZSPCDBTK9MCBS4QM

Comment (spacedmonkey) [reply src/wp-includes/theme.php]: What if the filter is added after the first time that `get_stylesheet_directory` is run. Like if it was run again init or plugins_loaded. I am pretty sure that `get_stylesheet_directory` is called very early in the bootstrap.

Comment (spacedmonkey) [reply src/wp-includes/theme.php]: Looks like this is first called in `wp_templating_constants` or on the `setup_theme` hook. If filter is hooked on `after_setup_theme`, `init` then it would not filter. I wonder if we can use `has_filter` here?

File: src/wp-includes/theme.php [hunk 13]
```diff
@@ -187,24 +187,33 @@ function get_stylesheet() {
...
-	 * Filters the stylesheet directory path for the active theme.
-	 *
-	 * @since 1.5.0
-	 *
-	 * @param string $stylesheet_dir Absolute path to the active theme.
-	 * @param string $stylesheet     Directory name of the active theme.
-	 * @param string $theme_root     Absolute path to themes directory.
-	 */
-	return apply_filters( 'stylesheet_directory', $stylesheet_dir, $stylesheet, $theme_root );
+	if ( null === $wp_stylesheet_path ) {
```
Comment (spacedmonkey) [inline src/wp-includes/theme.php:199]:
```suggestion
	if ( null === $wp_stylesheet_path || has_filter( 'stylesheet_directory' ) ) {
```

File: src/wp-includes/theme.php [hunk 14]
```diff
@@ -321,24 +330,33 @@ function get_template() {
...
-	 * Filters the active theme directory path.
-	 *
-	 * @since 1.5.0
-	 *
-	 * @param string $template_dir The path of the active theme directory.
-	 * @param string $template     Directory name of the active theme.
-	 * @param string $theme_root   Absolute path to the themes directory.
-	 */
-	return apply_filters( 'template_directory', $template_dir, $template, $theme_root );
+	if ( null === $wp_template_path ) {
```
Comment (spacedmonkey) [inline src/wp-includes/theme.php:349]:
```suggestion
	if ( null === $wp_template_path || has_filter( 'template_directory' ) ) {
```

File: tests/phpunit/tests/template.php (hunk 11 above)
Comment (spacedmonkey) [inline tests/phpunit/tests/template.php]: Instead of an array_filter, can we just use a foreach loop. It is cleaner.

File: src/wp-includes/theme.php (hunk 13 above)
Comment (felixarntz) [reply src/wp-includes/theme.php:199]: I like the overall idea behind it, but there are two issues with it:
* It defeats part of the purpose of the memoization, since now `has_filter()` will be called every time that the global is already set. I think we can solve that within the if clause, to only cache the value overall if there are no filters (which will be the case for most sites).
* If we check for the `stylesheet_directory` filter, we have to just as well check for the `stylesheet` and `theme_root` filters, which are also run within this clause. We have to consider the inner functions as well. I think we can do that though: Generally, all of these filters aren't very common to use, and in core itself they are only used in a few special places that doesn't impact general frontend performance.

I have pushed an alternate fix in https://github.com/WordPress/wordpress-develop/pull/5160/commits/1d3330e26ae0ceb742cf5fd137bfd078814c24d7.

File: src/wp-includes/theme.php (hunk 14 above)
Comment (felixarntz) [reply src/wp-includes/theme.php:349]: See my reply in https://github.com/WordPress/wordpress-develop/pull/5160#discussion_r1323519557.

File: tests/phpunit/tests/template.php (hunk 11 above)
Comment (felixarntz) [reply tests/phpunit/tests/template.php]: Fixed in https://github.com/WordPress/wordpress-develop/pull/5160/commits/29906b88ad0581b4b93a62cac987247314f08d6e

**Review by mukeshpanchal27 — APPROVED** [review]
Thank you, @felixarntz, for the update. It looks good from my end.

File: tests/phpunit/tests/theme.php [hunk 15]
```diff
@@ -282,23 +283,28 @@ public function test_switch_theme() {
 
 		for ( $i = 0; $i < 3; $i++ ) {
 			foreach ( $themes as $name => $theme ) {
+				// Skip invalid theme directory names (such as `block_theme-[0.4.0]`).
+				if ( ! preg_match( '/^[a-z0-9-]+$/', $theme['Stylesheet'] ) ) {
+					continue;
+				}
```
Comment (spacedmonkey) [inline tests/phpunit/tests/theme.php:289]: Why is this needed?

File: tests/phpunit/tests/theme.php [hunk 16]
```diff
@@ -309,19 +315,38 @@ public function test_switch_theme() {
 				$this->assertSame( $root_fs . '/' . get_template(), get_template_directory() );
 				$this->assertSame( $root_uri . '/' . get_template(), get_template_directory_uri() );
 
-				// get_query_template()
+				// Skip block themes for get_query_template() tests since this test is focused on classic templates.
+				if ( wp_is_block_theme() && current_theme_supports( 'block-templates' ) ) {
+					continue;
+				}
```
Comment (spacedmonkey) [inline tests/phpunit/tests/theme.php]: Why can't we test block themes here?

File: tests/phpunit/tests/theme.php [hunk 17]
```diff
@@ -854,6 +879,64 @@ public function test_block_theme_should_load_separate_core_block_assets_by_defau
...
 
+	/**
+	 * Tests that a theme in the custom test data theme directory is recognized.
+	 *
+	 * @ticket 18298
+	 */
+	public function test_theme_in_custom_theme_dir_is_valid() {
+		switch_theme( 'block-theme' );
+		$this->assertTrue( wp_get_theme()->exists() );
+	}
```
Comment (spacedmonkey) [inline tests/phpunit/tests/theme.php:890]: Why is this new test needed?

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Can we have a unit test for using `stylesheet` and `theme_root` filter in the unit tests. 

There is also some confusing things in the unit tests I would like unit tests. Otherwise the code is looking nearly good.

File: tests/phpunit/tests/theme.php (hunk 15 above)
Comment (felixarntz) [reply tests/phpunit/tests/theme.php:289]: @spacedmonkey This test is extremely old and honestly, poorly written as it includes lots of assertions for wildly different things (unit test standards back then were not what they are today).

We shouldn't have to completely update the tests to cover all the new requirements, new and more appropriate tests should be added for that.

This change here specifically is because some new test theme directories were added that include disallowed characters which make the tests below fail. That was never tested in this test, so we don't have to start now.

File: tests/phpunit/tests/theme.php (hunk 16 above)
Comment (felixarntz) [reply tests/phpunit/tests/theme.php]: @spacedmonkey See my reply in https://github.com/WordPress/wordpress-develop/pull/5160/files#r1331788580. We shouldn't add that here since this is a massive legacy test, written at a time where only classic themes existed. Most of what is tested here does not apply to block themes, and fixing this test to work for block themes is a waste of time.

File: tests/phpunit/tests/theme.php (hunk 17 above)
Comment (felixarntz) [reply tests/phpunit/tests/theme.php:890]: This is to cover a side effect from https://github.com/WordPress/wordpress-develop/pull/5160/files#diff-d40642a7652a1235b560872716da6cf1a4d406ddb8d55ba25761835d35c0a562R41: Since we now also cover the themes directory specific for tests here, it's worth ensuring that that works.

If it failed, all other tests below would fail for the same reason, so there should be a dedicated test to cover just the part of using a "non-default" themes directory.

File: tests/phpunit/tests/theme.php [hunk 18]
```diff
@@ -309,19 +315,38 @@ public function test_switch_theme() {
 				$this->assertSame( $root_fs . '/' . get_template(), get_template_directory() );
 				$this->assertSame( $root_uri . '/' . get_template(), get_template_directory_uri() );
 
-				// get_query_template()
+				// Skip block themes for get_query_template() tests since this test is focused on classic templates.
+				if ( wp_is_block_theme() && current_theme_supports( 'block-templates' ) ) {
```
Comment (spacedmonkey) [inline tests/phpunit/tests/theme.php]:
```suggestion
				if ( $theme->is_block_theme() && current_theme_supports( 'block-templates' ) ) {
```

**Comment by felixarntz** [comment]
@spacedmonkey I addressed your feedback above and added tests in https://github.com/WordPress/wordpress-develop/pull/5160/commits/963d54241baca732aa8df65b46d16fd3cc3edf49.

File: tests/phpunit/tests/theme.php (hunk 18 above)
Comment (felixarntz) [reply tests/phpunit/tests/theme.php]: That doesn't really make a difference. Also, we're calling `current_theme_supports()` and there's no `$theme->supports()` or similar available I think, so we might as well keep it as is for consistency.

Comment (spacedmonkey) [reply tests/phpunit/tests/theme.php]: This saves creating a new WP_Theme object, making the tests a little faster.

**Review by spacedmonkey — APPROVED** [review]

**Comment by felixarntz** [comment]
Committed in https://core.trac.wordpress.org/changeset/56635
